### PR TITLE
Add multi-valued fields support

### PR DESCRIPTION
**Pull request by astefan**
Draft WIP PR.

### CODE REVIEW

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/whitelist/InternalEqlScriptUtils.java [hunk 1]
```diff
@@ -33,6 +36,22 @@
     InternalEqlScriptUtils() {
     }
 
+    public static <T> Boolean multiValueDocValues(Map<String, ScriptDocValues<T>> doc, String fieldName, Predicate<T> script) {
+        if (doc.containsKey(fieldName)) {
+            ScriptDocValues<T> docValues = doc.get(fieldName);
+            if (docValues.isEmpty() == false) {
```
Comment (matriv) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/whitelist/InternalEqlScriptUtils.java:42]: Should be covered by the for loop and the `docValues.size()` check, no?
Or maybe a check if the `docValues` is `null` could be necessary (if that's a possibility).

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/planner/QueryTranslator.java [hunk 2]
```diff
@@ -65,11 +68,29 @@ public static Query toQuery(Expression e) {
...
+        int i = 0;
+        while (translation == null && i < QUERY_TRANSLATORS.size()) {
+            translation = QUERY_TRANSLATORS.get(i).translate(e, handler);
+            i++;
+        }
+        if (translation != null) {
+            if (translation instanceof ScriptQuery) {
+                // check the operators and the expressions involved in these operations so that all can be used
+                // in a doc-values multi-valued context
+                boolean multiValuedCompatible = false == e.anyMatch(exp -> {
```
Comment (matriv) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/planner/QueryTranslator.java:80]: Maybe consider extracting to a method.

**Review by matriv — COMMENTED** [review]
Looks good to me so far.

**Review by bpintea — COMMENTED** [review]
Approach looks good AFAI can tell. Left a few mostly superficial comments.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/gen/script/Scripts.java [hunk 3]
```diff
@@ -107,4 +112,95 @@ public static String classPackageAsPrefix(Class<?> function) {
...
+
+    /**
+     * This method replaces any .docValue(doc,params.%s) call with a variable.
+     * Each variable is then used in a {@code java.util.function.Predicate} to iterate over the doc_values in a Painless script
+     * 
+     * For example, a query of the form fieldA - fieldB > 0 that gets translated into the following Painless script
+     * {@code InternalQlScriptUtils.nullSafeFilter(InternalQlScriptUtils.gt(InternalQlScriptUtils.sub(
+     * InternalQlScriptUtils.docValue(doc,params.v0),InternalQlScriptUtils.docValue(doc,params.v1)),params.v2))}
+     * will become, after this method rewrite
+     * {@code InternalEqlScriptUtils.multiValueDocValues(doc,params.v0,X1 -> InternalEqlScriptUtils.multiValueDocValues(doc,params.v1,
```
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/gen/script/Scripts.java:124]: I don't think I'll merge that PR. It started loosing its purpose if we do carthesian product to handle multi-value fields.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/gen/script/Scripts.java [hunk 4]
```diff
@@ -107,4 +112,95 @@ public static String classPackageAsPrefix(Class<?> function) {
...
+                otherVars.add(params.get("v" + index));
+                index++;
+            } else {
+                newTemplate.append(token);
+            }
+        }
+
+        // iterate over the fields in reverse order and add a multiValueDocValues call for each
+        for(int i = fieldVars.size() - 1; i >= 0; i--) {
+            newTemplate.insert(0, "InternalEqlScriptUtils.multiValueDocValues(doc,params.%s,X" + i + " -> ");
```
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/gen/script/Scripts.java:160]: The tests complain if they find whitespaces.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/gen/script/Scripts.java [hunk 5]
```diff
@@ -107,4 +112,95 @@ public static String classPackageAsPrefix(Class<?> function) {
...
+        // the rest of variables come after
+        otherVars.forEach(v -> newParams.variable(v));
+
+        return new ScriptTemplate(newTemplate.toString(), newParams.build(), DataTypes.BOOLEAN);
+    }
+
+    /*
+     * Split a string given a regular expression into tokens. The list of tokens includes both the
+     * segments that matched the regex and also the segments that didn't.
+     * "fooxbarxbaz" split using the "x" regex will build an array like ["foo","x","bar","x","baz"]
```
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/gen/script/Scripts.java:176]: Maybe as an improvement later on.

**Pull request closed without merging**